### PR TITLE
Add function to read waagent.conf settings and condition to run it only on Linux

### PR DESCRIPTION
**Pull request by tatyana-kostromskaya**
Add function to read configs of waagent.conf and write it to separate diagnostics log file.
Add condition to run it only for Linux.

Tested this change manually:
1. Created Linux Azure VM
2. Configured agent with my changes on this VM.
3. Run several builds on Azure DevOps using this self-hosted agent and gathered logs.
4. Builds worked fine and logs contained new information

### CODE REVIEW

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 1]
```diff
@@ -153,6 +168,34 @@ private string GetCapabilitiesContent(Dictionary<string, string> capabilities)
             return builder.ToString();
         }
 
+        private string ParseWaagentContent(string configPath)
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:171]: Please add /// comment for new function

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 2]
```diff
@@ -105,6 +105,21 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
                 File.Copy(agentLogFile, destination);
             }
 
+            // Read and add to logs waagent.conf settings on Linux
+            if (PlatformUtil.RunningOnLinux)
+            {
+                executionContext.Debug("Creating waagent file.");
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:111]: Maybe "Dumping of waagent.conf file"?

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 3]
```diff
@@ -105,6 +105,21 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
...
+                string waagentFile = Path.Combine(supportFilesFolder, "waagentConf.txt");
+
+                string configFileName = "waagent.conf";
+                string filePath = WhichUtil.Which(configFileName);
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:115]: Do we need to check Path variable for waagent.conf file path?

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 4]
```diff
@@ -105,6 +105,21 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
...
+                if(string.IsNullOrWhiteSpace(filePath)){
+                    filePath = Directory.GetFiles("/etc", configFileName).FirstOrDefault();
+                }
+                string waagentContent = !string.IsNullOrEmpty(filePath) ? ParseWaagentContent(filePath) : "waagent.conf file is not found";
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:119]: Nit: waagentContent is not really exact here - "waagent.conf file is not found" is the error message, but not the file content.

File: src/Agent.Worker/DiagnosticLogManager.cs (hunk 2 above)
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:111]: Please add fallback - so if any error happens during new logic, it won't crush agent, but will print error message in logs.

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 5]
```diff
@@ -153,6 +168,34 @@ private string GetCapabilitiesContent(Dictionary<string, string> capabilities)
             return builder.ToString();
         }
 
+        private string ParseWaagentContent(string configPath)
+        {
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:172]: Do we really need this parsing?

Comment (tatyana-kostromskaya) [reply src/Agent.Worker/DiagnosticLogManager.cs:172]: I looked at options in waagent.conf and I did not find any PII data there, but this parsing allows to skip comments. I am not sure if comments are needed in the dump and if the may contain any sensitive data.

If we want to leave comments in the dump, then we can just dump the whole file without parsing.

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 6]
```diff
@@ -105,6 +105,37 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
...
+
+                        File.AppendAllText(waagentDumpFile, "waagent.conf settings");
+                        File.AppendAllText(waagentDumpFile, Environment.NewLine);
+                        File.AppendAllText(waagentDumpFile, waagentContent);
+                        
+                        executionContext.Debug("Dumping waagent.conf file is completed.");
+                    }
+                    else
+                    {
+                        executionContext.Warning("waagent.conf file wasn,t found. Dumping was not done.");
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:130]: Nit: was't

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 7]
```diff
@@ -105,6 +105,37 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
                 File.Copy(agentLogFile, destination);
             }
 
+            // Read and add to logs waagent.conf settings on Linux
+            if (PlatformUtil.RunningOnLinux)
+            {
+                executionContext.Debug("Dumping of waagent.conf file");
+                string waagentDumpFile = Path.Combine(supportFilesFolder, "waagentConf.txt");
+
+                string configFileName = "waagent.conf";
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:114]: Do you know if this file could be located by some custom path?

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 8]
```diff
@@ -105,6 +105,37 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
...
+                        executionContext.Debug("Dumping waagent.conf file is completed.");
+                    }
+                    else
+                    {
+                        executionContext.Warning("waagent.conf file wasn,t found. Dumping was not done.");
+                    }
+                }
+                catch
+                {
+                    executionContext.Warning("Dumping of waagent.conf was not completed successfully.");
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:135]: Please output the error message also

File: src/Agent.Worker/DiagnosticLogManager.cs (hunk 7 above)
Comment (tatyana-kostromskaya) [reply src/Agent.Worker/DiagnosticLogManager.cs:114]: Right now waagent support only this standard path. Right now I don't see any ticket in https://github.com/Azure/WALinuxAgent with requests to support custom paths

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 9]
```diff
@@ -105,6 +105,38 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
...
+                    string filePath = Directory.GetFiles("/etc", configFileName).FirstOrDefault();
+                    if (!string.IsNullOrWhiteSpace(filePath))
+                    {
+                        string waagentContent = File.ReadAllText(filePath);
+
+                        File.AppendAllText(waagentDumpFile, "waagent.conf settings");
+                        File.AppendAllText(waagentDumpFile, Environment.NewLine);
+                        File.AppendAllText(waagentDumpFile, waagentContent);
+                        
+                        executionContext.Debug("Dumping waagent.conf file is completed.");
```
Comment (DaniilShmelev) [inline src/Agent.Worker/DiagnosticLogManager.cs:126]: Maybe we should also include this file in the diagnostics logs archive? @anatolybolshakov what do you think?

Comment (anatolybolshakov) [reply src/Agent.Worker/DiagnosticLogManager.cs:126]: @DaniilShmelev not really sure - I would leave it in the same location as environment.txt and capabilities.txt

**Review by anatolybolshakov — APPROVED** [review]

**Review by DaniilShmelev — APPROVED** [review]

**Pull request merged**